### ISSUE TITLE
Support for JS Component Types

### ISSUE THREAD

**Issue by imcharsi** [japgolly/scalajs-react#157]
Hi.

At first, My English is poor. I'm sorry about that.

I think that it would be better if we can use below method.
```
createElement(component: js.Dynamic, props: js.Object, children: ReactNode*): ReactComponentU_
```

In javascript, we can write below code with Material-UI.
```
React.createElement(mui.TextField, {someProp...}, someChildren...);
```

[Here is chandu0101's example.](https://github.com/chandu0101/scalajs-react-components/blob/master/core/src/main/scala/chandu0101/scalajs/react/components/materialui/package.scala#L42)
```
    val TextField: js.Dynamic = js.native
```

And, [Here.](https://github.com/chandu0101/scalajs-react-components/blob/master/core/src/main/scala/chandu0101/scalajs/react/components/materialui/MuiTextField.scala#L69)
```
    val f = React.asInstanceOf[js.Dynamic].createFactory(Mui.TextField)
    f(p).asInstanceOf[ReactComponentU_]
```

So, If we can use above method, we can write below code.
```
React.createElement(Mui.TextField, p)
```

Thanks.

**Comment by chandu0101** [comment]
+1

**Comment by japgolly** [comment]
Using `js.Dynamic` is lazy. Should be using JS types like `ReactComponentC_`, `ReactComponentU_`, `ReactComponentM_` for JS components. Is there a type missing for what gets fed into `createElement`?

With Scala, where … = `ReactComponent`, it's
`…B` ⇒  `…Spec` ⇒ `…Type` ⇒ `…C` ⇒ `…U` ⇒ `…M`.

I don't mind a more JS-friendly `createElement` being added but first you'll need to replace `js.Dynamic` with something more appropriate.

(I notice that Scala components' internal JS component ctor is typed `ReactComponentCU`. That seems a bit weird - either I'm reading too quickly now or I was accidentally inconsistent in the past.  Either way I wonder if that type might overlap with what you need here.)

**Comment by imcharsi** [comment]
Hi.

Thanks for reply.

At first, If below my suggestion is not your intended usage, please tell me about your example usage.

So, [with below API](https://github.com/japgolly/scalajs-react/blob/master/core/src/main/scala/japgolly/scalajs/react/React.scala#L18),

```
  def createFactory[P,S,B,N <: TopNode](t: ReactComponentType[P,S,B,N]): ReactComponentCU[P,S,B,N] = js.native
```

I tried below code.

```
@JSName("mui.TextField")
object Xxx extends ReactComponentType[CustomProperty, Unit, Unit, HTMLElement]

React.createFactory(Xxx)(WrapObj())
```

When I tried above code, I could see the component in web browser. But, When I tried below code,

```
React.createFactory(Xxx)(WrapObj(CustomProperty(disabled = true, ...)), someChildren...)
```

I couldn't set any properties.

So, I tried below code, But It was not compiled.

```
React.createFactory(Xxx)(CustomProperty(disabled = true, ...), someChildren...)
```

[Below is related API](https://github.com/japgolly/scalajs-react/blob/master/core/src/main/scala/japgolly/scalajs/react/React.scala#L139).

```
trait ReactComponentCU[Props, State, +Backend, +Node <: TopNode] extends ReactComponentC_ with ReactComponentTypeAuxJ[Props, State, Backend, Node] {
  def apply(props: WrapObj[Props], children: ReactNode*): ReactComponentU[Props, State, Backend, Node] = js.native
}
```

The method always receives property as WrapObj. So, I suggest below api.

```
  def apply(props: Props, children: ReactNode*): ReactComponentU[Props, State, Backend, Node] = js.native
```
  
Changed api is like below.

```
trait ReactComponentCU[Props, State, +Backend, +Node <: TopNode] extends ReactComponentC_ with ReactComponentTypeAuxJ[Props, State, Backend, Node] {
  def apply(props: WrapObj[Props], children: ReactNode*): ReactComponentU[Props, State, Backend, Node] = js.native
  def apply(props: Props, children: ReactNode*): ReactComponentU[Props, State, Backend, Node] = js.native // <-- added
}
```

After I cloned the source,
```
  git clone https://github.com/japgolly/scalajs-react.git
```
and added the method and build the source, below code works.

```
React.createFactory(Xxx)(CustomProperty(disabled = true, ...), someChildren...)
```

Thanks.

**Comment by japgolly** [comment]
Hi @imcharsi, I fear we have a misunderstanding. Your goal is you want the `React` object to be easier to use with pure JS components, yes? Firstly you cannot use the Scala component types because they are specialised for Scala components. This means you cannot use the `ReactComponentType`.

I don't use pure-JS components - Scala as always been my priority so I've never put the work in to support nice JS-component usage. I'm happy for someone to make it work but I don't have time to do it. Thus, my suggestion is that to support your use case, you may (probably) need to create some new types.

I showed you the pipeline for creating Scala components above (…B ⇒ …Spec ⇒ …Type ⇒ …C ⇒ …U ⇒ …M) - now I think for your case you need a similar pipeline but for JS components. Some of the types already exist for JS; they end in underscore like `ReactComponentU_` and `ReactComponentM_`. My idea is maybe you also need a `ReactComponentType_` and/or a  `ReactComponentSpec_` or whatever. When you have the types that JS needs then a new `React.createFactory` method could be added that uses only the JS types.

**Comment by imcharsi** [comment]
Hi.

```
Your goal is you want the React object to be easier to use with pure JS components, yes?
```
Yes.
```
Firstly you cannot use the Scala component types because they are specialised for Scala components.
```
I got it.
```
When you have the types that JS needs then a new React.createFactory method could be added that uses only the JS types.
```
Ok.

Your reply is very helpful. Thanks.

**Comment by japgolly** [comment]
Merged #158
